### PR TITLE
rqt controller manager ros 2 port

### PR DESCRIPTION
**Pull request by ijnek**
This is a port of rqt_controller_manager from ROS 1 to ROS 2, but with some modifications listed below that were necessary due to fundamental changes in the ros2 control architecture. This PR probably still requires some cleaning up, but I'd appreciate any early reviews and testers!

  - states of the controllers have changed
    - not loaded
    - unconfigured
    - inactive
    - active
    - finalized
  - the controller namespace has been replaced with controller node name
  - parameters are obtained directly from controller manager using get_parameter service rather than accessing parameter server
  - replaces manual service calls (configure_controller, load_controller, etc.) with helper methods from ``controller_manager_services``
  
The commits aren't well organized (because I was learning the ros2 control structure as I was going), so I don't recommend reviewing per commit.
  
ps.
The ROS1 version uses BSD. This port has modified it to Apache 2.0 because that was the only way I could get the tests passing, and it seems like everything else has been changed to apache 2.0 during hte ROS2 port.

Resolves: #609

### CODE REVIEW

**Comment by destogl** [comment]
This looks great!! Thanks! I test it and it works.

It would be great if you can support all the transitions we have in the lifecycle (except to error, finalized and unknown).

Also in the next PR we can add another list under with list of hardware and their lifecycle! This would make examples much more visual.

File: rqt_controller_manager/rqt_controller_manager/update_combo.py
Comment (erickisos) [inline rqt_controller_manager/rqt_controller_manager/update_combo.py:56]: Just random question, do we expect to have repeated values?

Otherwise, maybe we can check with something like this 
```suggestion
    return set(a) == set(b)
```

**Review by erickisos — APPROVED** [review]
Looks stunning! 👏

**Comment by ijnek** [comment]
Could you help me write out all the states and transitions? The controller lifecycle seems slightly different to the [ROS 2 Node Lifecycle](https://design.ros2.org/articles/node_lifecycle.html).

Refering to
https://github.com/ros-controls/ros2_control/blob/94d1e9a9f07c92094992d26c44ac8d9e0a3c9103/controller_manager_msgs/msg/ControllerState.msg#L2

the states available are:

* ``unconfigured``
* ``configured``
* ``active``
* ``inactive`` (_same as ``configured``?_)

And I believe the possible transitions are:

* Load (not loaded -> ``unconfigured``)
* Configure (``unconfigured`` -> ``inactive``)
* Activate (``inactive`` -> ``activate``)
* Deactivate (``active`` -> ``inactive``)
* Unload (``inactive`` -> not loaded)

This is excluding composed transitions. (eg. Kill (``active`` -> ``inactive`` -> not loaded) which can be broken down into Deactive + Unload)

File: rqt_controller_manager/rqt_controller_manager/update_combo.py
Comment (ijnek) [reply rqt_controller_manager/rqt_controller_manager/update_combo.py]: I don't think having repeated values is possible, so I'm going to commit your suggestion.

**Comment by destogl** [comment]
> Could you help me write out all the states and transitions? The controller lifecycle seems slightly different to the [ROS 2 Node Lifecycle](https://design.ros2.org/articles/node_lifecycle.html).

The states are actually the same (at least this is the target): https://github.com/ros-controls/ros2_control/issues/551#issuecomment-947174795

`configured` is the old nomenclature that we are getting rid off, but we try to make the process not so disruptive.

Your lists are looking correct!

File: rqt_controller_manager/rqt_controller_manager/controller_manager.py
Comment (erickisos) [inline rqt_controller_manager/rqt_controller_manager/controller_manager.py:331]: Perhaps, we can reduce the complexity of this function with something like this
```suggestion
    def headerData(self, col, orientation, role):
        if orientation != Qt.Horizontal or role != Qt.DisplayRole:
            return None
        if col == 0:
            return 'controller'
        elif col == 1:
            return 'state'
```

The weird thing is that I don't know if we can have another value for `col`, otherwise we can do a `return 'state' if col == 1 else 'controller'`

File: rqt_controller_manager/rqt_controller_manager/controller_manager.py
Comment (erickisos) [inline rqt_controller_manager/rqt_controller_manager/controller_manager.py:346]:
```suggestion
                return ctrl.state or 'not loaded'
```

File: rqt_controller_manager/rqt_controller_manager/controller_manager.py
Comment (erickisos) [inline rqt_controller_manager/rqt_controller_manager/controller_manager.py:353]: The next reduction assumes that if we don't have `ctrl.state` in `self._icons` we are going to retrieve a `None` without extra issues.
```suggestion
        if role == Qt.DecorationRole and index.column() == 0:
            return self._icons.get(ctrl.state)
```

File: rqt_controller_manager/rqt_controller_manager/controller_manager.py
Comment (erickisos) [inline rqt_controller_manager/rqt_controller_manager/controller_manager.py:359]:
```suggestion
        if role == Qt.FontRole and index.column() == 0:
            bf = QFont()
            bf.setBold(True)
            return bf
```

File: rqt_controller_manager/rqt_controller_manager/controller_manager.py
Comment (erickisos) [inline rqt_controller_manager/rqt_controller_manager/controller_manager.py:363]:
```suggestion
        if role == Qt.TextAlignmentRole and index.column() == 1:
            return Qt.AlignCenter
```

File: rqt_controller_manager/rqt_controller_manager/controller_manager.py
Comment (erickisos) [inline rqt_controller_manager/rqt_controller_manager/controller_manager.py:400]:
```suggestion
    return response.values[0].string_value if response.values else ''
```

File: rqt_controller_manager/rqt_controller_manager/controller_manager.py
Comment (erickisos) [inline rqt_controller_manager/rqt_controller_manager/controller_manager.py:418]:
```suggestion
    return [
        name.rstrip('list_controllers').rstrip('/')
        for name, _ in get_service_names_and_types(node=node)
        if name.endswith('list_controllers')
    ]
```

**Review by erickisos — APPROVED** [review]
I left some other suggestions, however, this looks great.

Idk if we should add some extra tests for the controller manager itself.

File: rqt_controller_manager/package.xml
Comment (destogl) [inline rqt_controller_manager/package.xml:18]: @bmagyar should we clean this up for the ros2 version?

File: rqt_controller_manager/package.xml
Comment (destogl) [inline rqt_controller_manager/package.xml:7]:
```suggestion
  <maintainer email="kenjibrameld@gmail.com">ijnek</maintainer>
  <maintainer email="bence.magyar.robotics@gmail.com">Bence Magyar</maintainer>
  <maintainer email="denis@stoglrobotics.de">Denis Štogl</maintainer>
```

**Comment by destogl** [comment]
When backporting to humble the commit `Updated to work on rolling.` should be removed.

**Comment by ijnek** [comment]
Thanks @destogl and @erickisos for the reviews! I wasn't quite finding the time to address the suggestions, so I appreciate you going through them @destogl, thanks .

**Comment by destogl** [comment]
@ijnek This is currently working good, but I see few improvements we have to make. Those can also be done after the merge.

Do you think that we should merge it, and then we can address other issues in separate PRs? Maybe more people can join the party then also.

the state you currently achieved it's great and comparable to the ROS1 functionalities so for me it's OK to merge this.

@bmagyar what do you think?

**Comment by destogl** [comment]
We should also address the following API changes in `ros2param` package:

- [ ] https://github.com/ros2/ros2cli/pull/716
- [ ] https://github.com/ros2/rclpy/pull/945/files

This means, we should probably use service client from `rclpy` and not `ros2param` package. At least in some cases.

**Comment by ijnek** [comment]
As long as they're tracked in an issue, I'd prefer separate issues so others can perhaps take-on the tasks.

File: rqt_controller_manager/package.xml
Comment (bmagyar) [reply rqt_controller_manager/package.xml:18]: I think we can also leave it as-is.

File: rqt_controller_manager/setup.py
Comment (bmagyar) [inline rqt_controller_manager/setup.py:20]: Are you staying around to maintain the package? More than welcome to but I'd you'd rather not you can also change the contact info in here for mine.

File: rqt_controller_manager/setup.py
Comment (bmagyar) [inline rqt_controller_manager/setup.py:22]: Apache 2

File: rqt_controller_manager/test/test_copyright.py
Comment (bmagyar) [inline rqt_controller_manager/test/test_copyright.py:1]: I believe this is part of our CI now, no need to copy them around

File: rqt_controller_manager/test/test_flake8.py
Comment (bmagyar) [inline rqt_controller_manager/test/test_flake8.py:2]: I believe this is part of our CI now, no need to copy them around

File: rqt_controller_manager/test/test_pep257.py
Comment (bmagyar) [inline rqt_controller_manager/test/test_pep257.py:2]: I believe this is part of our CI now, no need to copy them around

File: rqt_controller_manager/setup.py
Comment (ijnek) [reply rqt_controller_manager/setup.py:20]: I put my name in to ensure I don't push work on to you, but I believe things are going to be easier if you are the maintainer. If you're OK with me putting your name down, I'd appreciate that.

Comment (bmagyar) [reply rqt_controller_manager/setup.py:20]: Yeah that's my job here ;)

**Comment by ijnek** [comment]
@bmagyar addressed everything

**Comment by ijnek** [comment]
Is there anything still blocking this from getting merged?

**Review by bmagyar — APPROVED** [review]

**Comment by bmagyar** [comment]
@mergifyio backport to humble

**Comment by ijnek** [comment]
Thanks @bmagyar and @destogl !

**Pull request merged**
